apps/builder/app/shared/style-object-model.ts

```diff
@@ -0,0 +1,155 @@
+};
+
+type Property = keyof typeof properties;
+
+const getCascadedValue = ({
+  model,
+  instanceId,
+  property,
+}: {
+  model: StyleObjectModel;
+  instanceId: Instance["id"];
+  property: Property;
+}) => {
+  // https://drafts.csswg.org/css-cascade-5/#declared
+  const declaredValues = [];
+  const styleSourceIds = model.styleSourcesByInstanceId.get(instanceId);
+  if (styleSourceIds) {
+    for (const styleSourceId of styleSourceIds) {
+      const style = model.styleByStyleSourceId.get(styleSourceId);
+      const styleDecl = style?.get(property);
+      if (styleDecl) {
+        declaredValues.push(styleDecl.value);
+      }
+    }
+  }
+
+  // https://drafts.csswg.org/css-cascade-5/#cascaded
+  const cascadedValue = declaredValues.at(-1);
+  return { cascadedValue };
+};
+
+const isKeyword = (styleValue: undefined | StyleValue, keyword: string) =>
+  styleValue?.type === "keyword" && styleValue.value === keyword;
+
+/**
+ * follow value processing specification
+ * https://drafts.csswg.org/css-cascade-5/#value-stages
+ *
+ * @todo
+ * - current color
+ * - custom property
+ * - html
+ * - preset
+ * - cascaded
+ * - selected style source
+ * - selected state
+ * - breakpoints
+ *
+ */
+export const getComputedStyleDecl = ({
+  model,
+  styleSelector,
+  property,
+}: {
+  model: StyleObjectModel;
+  styleSelector: StyleSelector;
+  property: Property;
+}) => {
+  const { instanceSelector } = styleSelector;
+  const propertyData = properties[property];
+  const inherited = propertyData.inherited;
+  const initialValue = propertyData.initial;
+  let computedValue: StyleValue = initialValue;
+
+  // start computing from the root
+  for (const instanceId of instanceSelector.slice().reverse()) {
+    // https://drafts.csswg.org/css-cascade-5/#inheriting
+    const inheritedValue: StyleValue = computedValue;
+
+    // https://drafts.csswg.org/css-cascade-5/#cascaded
+    const { cascadedValue } = getCascadedValue({ model, instanceId, property });
+
+    // https://drafts.csswg.org/css-cascade-5/#specified
+    let specifiedValue: StyleValue;
+    // explicit defaulting
+    // https://drafts.csswg.org/css-cascade-5/#defaulting-keywords
+    if (isKeyword(cascadedValue, "initial")) {
+      specifiedValue = initialValue;
+    } else if (isKeyword(cascadedValue, "inherit")) {
+      specifiedValue = inheritedValue;
+    } else if (isKeyword(cascadedValue, "unset")) {
+      if (inherited) {
+        specifiedValue = inheritedValue;
+      } else {
+        specifiedValue = initialValue;
+      }
+    } else if (cascadedValue) {
+      specifiedValue = cascadedValue;
+    }
+    // defaulting https://drafts.csswg.org/css-cascade-5/#defaulting
+    else if (inherited) {
+      specifiedValue = inheritedValue;
+    } else {
+      specifiedValue = initialValue;
+    }
+
+    // https://drafts.csswg.org/css-cascade-5/#computed
+    computedValue = specifiedValue;
+  }
+
+  return { computedValue };
```

Comment:
same question here, I didn't run ts compiler but from the look of it I see that it expects:

```
{
  model: StyleObjectModel;
  styleSelector: StyleSelector;
  property: Property;
}
```

Reply:
this object is input, the output will get other features later